lib/src/code_field/code_controller.dart

```diff
@@ -334,6 +334,151 @@ class CodeController extends TextEditingController {
     );
   }
 
+  void outdentSelection() {
+    final tabSpaces = params.tabSpaces;
+    if (selection.start == -1 || selection.end == -1) {
+      return;
+    }
+
+    modifySelectedLines(
+      (line) {
+        if (line == '\n') {
+          return line;
+        }
+
+        if (line.length < tabSpaces) {
+          return line.trimLeft();
+        }
+
+        final subStr = line.substring(0, tabSpaces);
+        if (subStr == ' ' * tabSpaces) {
+          return line.substring(tabSpaces, line.length);
+        }
+        return line.trimLeft();
+      },
+      shouldSustainSelection: true,
+    );
+  }
+
+  void indentSelection() {
+    final tabSpaces = params.tabSpaces;
+    final tab = ' ' * tabSpaces;
+    final lines = _code.lines.lines;
+    if (selection.start == -1 || selection.end == -1) {
+      return;
+    }
+
+    if (selection.isCollapsed) {
+      final cursor = _code.hiddenRanges.recoverPosition(
+        selection.start,
+        placeHiddenRanges: TextAffinity.downstream,
+      );
+      final lineIndex = _code.lines.characterIndexToLineIndex(cursor);
+      final columnIndex = cursor - lines[lineIndex].textRange.start;
+      final insert = ' ' * (tabSpaces - (columnIndex % tabSpaces).abs());
+      insertStr(insert);
+      return;
+    }
+
+    modifySelectedLines(
+      (line) {
+        if (line == '\n') {
+          return line;
+        }
+        return tab + line;
+      },
+      shouldSustainSelection: true,
+    );
+  }
+
+  /// Modify the lines, that are currently affected by selection.
+  /// All lines, except the last one, contain '\n' symbol at the end.
```

Comment:
Is this an input requirement or an output result? What if the last selected line has its `\n` selected?

Reply:
Moved this description to appropriate place

Reply:
> What if the last selected line has its \n selected?
Next line is also considered to be selected.